pkg/ioutils/copy.go

```diff
@@ -0,0 +1,57 @@
+package ioutils
+
+import (
+	"context"
+	"io"
+)
+
+// CopyCtx copies from src to dst until either EOF is reached on src or a context is cancelled.
+// The writer is not closed when the context is cancelled.
+//
+// After CopyCtx exits due to context cancellation, the goroutine that performed
+// the copy may still be running if either the reader or writer blocks.
+func CopyCtx(ctx context.Context, dst io.Writer, src io.Reader) (n int64, err error) {
+	copyDone := make(chan struct{})
+
+	src = &readerCtx{ctx: ctx, r: src}
+
+	go func() {
+		n, err = io.Copy(dst, src)
+		close(copyDone)
+	}()
+
+	select {
+	case <-ctx.Done():
+		return -1, ctx.Err()
+	case <-copyDone:
+	}
+
+	return n, err
+}
+
+type readerCtx struct {
+	ctx context.Context
+	r   io.Reader
+}
+
+// NewCtxReader wraps the given reader with a reader that doesn't proceed with
+// reading if the context is done.
+//
+// Note: Read will still block if the underlying reader blocks.
+func NewCtxReader(ctx context.Context, r io.Reader) io.Reader {
```

Comment:
I was curious as well if we checked if the Go team is considering something along these lines (and/or if there's proposals), and if we in that case could align to whatever design may be worked on.

I did find this proposal, but haven't read in-depth if there's any progress / good proposal on design there;

- https://github.com/golang/go/issues/20280